src/OpenApi/src/Extensions/OpenApiServiceCollectionExtensions.cs

```diff
@@ -57,13 +57,35 @@ public static IServiceCollection AddOpenApi(this IServiceCollection services, st
         ArgumentNullException.ThrowIfNull(services);
         ArgumentNullException.ThrowIfNull(configureOptions);
 
-        services.AddOpenApiCore(documentName);
-        services.Configure<OpenApiOptions>(documentName, options =>
+        // We need to store the document name in a case-insensitive manner
+        // to support case-insensitive document name resolution.
+        // Keyed Services are case-sensitive by default, which doesn't work well for document names in ASP.NET Core
+        // as routing in ASP.NET Core is case-insensitive by default.
+        var lowercasedDocumentName = documentName.ToLowerInvariant();
+
+        AddOpenApiCore(services, lowercasedDocumentName);
+        services.Configure<OpenApiOptions>(lowercasedDocumentName, options =>
         {
-            options.DocumentName = documentName;
+            options.DocumentName = lowercasedDocumentName;
             configureOptions(options);
         });
         return services;
+
+        // The reason this method is a local function is to prevent case-sensitive document names being passed into this method (from other methods) in the future.
```

Comment:
Is this necessary if the method was previously private and only accessible from methods in this class?

Reply:
It's not 100% required. It's a bit of defensive programming as explained in the comment. :) 

Keeping it a private method _might_ introduce scenarios in the future where non-lowercased input is passed into it, which would reintroduce the bug. 

With the code written this way, it's less likely.

Let me know if you'd like me to revert this! :)

Reply:
I have reset it! Let me know if there's anything else I can still do! Otherwise, feel free to merge!